File: src/chapters/numerical-calculus/content/code/composite.jl
function composite(f, a, b, n = 10)
    isodd(n) && (n += 1)                 # Simpson needs even n
    h = (b - a) / n
    T = (f(a) + f(b)) / 2
    S = f(a) + f(b)
    for i in 1:n-1
        yi = f(a + i*h)
        T += yi
        S += (isodd(i) ? 4 : 2) * yi
    end
    return h*T, h/3*S
end

T, S = composite(exp, 0.0, 1.0, 10)
println("trapezoid = ", T)
println("Simpson   = ", S)
# -> trapezoid ~ 1.7197135, Simpson ~ 1.7182828; exact e-1 = 1.7182818

File: src/chapters/numerical-calculus/content/code/composite.m
function [T, S] = composite(f, a, b, n)
% COMPOSITE  Composite trapezoidal (T) and Simpson (S) on [a,b], n subintervals.
    if nargin < 4, n = 10; end
    if mod(n, 2), n = n + 1; end          % Simpson needs even n
    h = (b - a) / n;
    x = a + (0:n) * h;
    y = arrayfun(f, x);
    T = h * (y(1)/2 + sum(y(2:end-1)) + y(end)/2);
    S = h/3 * (y(1) + y(end) + 4*sum(y(2:2:end-1)) + 2*sum(y(3:2:end-1)));
end

% --- Demo ---
[T, S] = composite(@exp, 0, 1, 10);
fprintf('trapezoid ~ %.7f\n', T);
fprintf('Simpson   ~ %.7f\n', S);
% -> trapezoid ~ 1.7197135, Simpson ~ 1.7182828; exact e-1 = 1.7182818
